src/lib/update-lock.ts

```diff
@@ -152,20 +103,41 @@ export function lock<T extends unknown>(
 								return Bluebird.mapSeries(
 									lockFilesOnHost(appId, serviceName),
 									(tmpLockName) => {
-										return Bluebird.try(() => {
-											if (force || lockOverride) {
-												return lockFile.unlockAsync(tmpLockName);
-											}
-										})
-											.then(() => lockFile.lockAsync(tmpLockName))
-											.then(() => {
-												locksTaken[tmpLockName] = true;
+										return (
+											Bluebird.try(() => {
+												if (force || lockOverride) {
+													return lockfile.unlock(tmpLockName);
+												}
 											})
-											.catchReturn(ENOENT, undefined);
+												.then(() => {
+													return lockfile.lock(tmpLockName);
+												})
+												// If lockfile exists, dispose of writeLock and throw a user-friendly error
+												.catch(lockfile.LockfileExistsError, () => {
+													return dispose(release).throw(
```

Comment:
Why is the `dispose` needed here if there is a disposer set at the end of the promise chain?

Reply:
The disposer is a meta method that only calls the function provided to `.disposer` when invoked in the context of `Bluebird.using`. However, when an error is thrown in `takeTheLock`, `Bluebird.using` is never reached in updateLock.lock.. so an additional call to `dispose(release)` is needed outside of `disposer(dispose)`. I changed it to be a bit more concise though, by attaching a `.finally(() => dispose(release)` to the end of the inner `Bluebird.mapSeries` like so:
```
return Bluebird.mapSeries(
	lockFilesOnHost(appId, serviceName),
	(tmpLockName) => {
        // abbreviated for brevity 
	}
).finally(() => dispose(release));
```

it'd be nice to move away from this nested structure though, because it's not easy to follow. The only thing preventing this is that there are some Bluebird-specific features used here that aren't present with native Promises..